Basic implementation for `basicUnsafeMove`

dmalkr opened the pull request:
Implementation for `basicUnsafeMove` and my own test for this -- it just goes through all the offset/stride/length options of source/target vectors. Arbitrary tests from `vector` package still not adopted.

Please review.

dmalkr committed:
Basic implementation for `basicUnsafeMove`

dmalkr committed:
test for intersected

dmalkr committed:
implementation for intersected vectors

dmalkr committed:
Refactor & cleaning

dmalkr committed:
Fix to new version

Shimuuar commented on vecvec-lapack/Vecvec/LAPACK/Internal/Compat.hs:
```diff
@@ -33,3 +36,13 @@ updPtr f (ForeignPtr p c) = case f (Ptr p) of { Ptr q -> ForeignPtr q c }
+-- | Distance in elements between two pointers. Supplement for `advancePtr`.
+--
+distancePtr :: forall a . Storable a => Ptr a -> Ptr a -> Int
+-- distancePtr ptrFrom ptrTo = (ptrTo `minusPtr` ptrFrom) `div` (sizeOf (undefined :: a))
+distancePtr = distancePtr' undefined -- TODO WTF?? why ambigous??
+ where
+    distancePtr' :: Storable a => a -> Ptr a -> Ptr a -> Int
+    distancePtr' x ptrFrom ptrTo = (ptrTo `minusPtr` ptrFrom) `div` sizeOf x
```
Without `ScopedTypeVariables` GHC will think that `a` is some unrelated type variable.

dmalkr replied:
Weird. I thought I tried `ScopedTypeVariables` :)

Shimuuar commented on vecvec-lapack/tests/TST/Memory.hs:
```diff
@@ -0,0 +1,59 @@
+
+
+tests :: TestTree
+tests = testGroup "memory"
```
Test could be simplified. Specification for move is quite simple: If vectors overlap, the copying is performed as if the source vector were copied to a temporary vector and then the temporary vector was copied to the target vector. So two program should be equivalent:

```haskell
carrier <- MVG.generateM carrierLength id
let source =  slice ... carrier
let target =  slice ... carrier
MVG.move target source
VG.unsafeFreeze carrier
```

```haskell
carrier <- MVG.generateM carrierLength id
let source =  slice ... carrier
let target =  slice ... carrier
buf <- MVG.clone source
MVG.copy target buf
VG.unsafeFreeze carrier
```

dmalkr replied:
Done

Shimuuar commented on vecvec-lapack/Vecvec/LAPACK/Internal/Vector/Mutable.hs:
```diff
@@ -212,17 +212,59 @@ instance VS.Storable a => MVG.MVector MVec a where
+        copy intersectPosition (-1)
+        copy (intersectPosition + 1) len
+    else
+      case () of
```
This is purely syntactic. `if ... then ... else case () of ...` could be replaced with `MultiWayIf`

dmalkr replied:
Done

Shimuuar reviewed (commented):
Overall  everything looks good. Except test could be simplified

dmalkr committed:
Using `MultiWayIf`

dmalkr committed:
Using `ScopedTypeVariables`

dmalkr committed:
Simplificate tests

dmalkr commented on vecvec-lapack/tests/TST/Memory.hs:
```diff
@@ -0,0 +1,56 @@
+                      tmp <- MVG.clone source2
+                      MVG.basicUnsafeCopy target2 tmp
+                      --
+                      (,,,) <$> VG.unsafeFreeze target1 <*> VG.unsafeFreeze target2 <*> VG.unsafeFreeze carrier1 <*> VG.unsafeFreeze carrier2
```
This use of the applicative and `unsafeFreeze` is safe, right?

Shimuuar committed:
Some cleanup of tests

 - Use safe variant for all functions. We don't try squeeze more performance so
   extra checks and copying won't hurt.
 - Factor out slice spec to a common definition
